tar/untar_stream.ts

```diff
@@ -185,43 +187,50 @@ export class UntarStream
     >();
     this.#readable = ReadableStream.from(this.#untar());
     this.#writable = writable;
+    this.#reader = readable.pipeThrough(new FixedChunkStream(512)).getReader();
+  }
 
-    this.#gen = async function* () {
-      const buffer: Uint8Array[] = [];
-      for await (
-        const chunk of readable.pipeThrough(new FixedChunkStream(512))
-      ) {
-        if (chunk.length !== 512) {
-          throw new RangeError(
-            `Cannot extract the tar archive: The tarball chunk has an unexpected number of bytes (${chunk.length})`,
-          );
-        }
+  async #read(): Promise<Uint8Array | undefined> {
+    const { done, value } = await this.#reader.read();
+    if (done) return undefined;
+    if (value.length !== 512) {
+      throw new RangeError(
+        `Cannot extract the tar archive: The tarball chunk has an unexpected number of bytes (${value.length})`,
+      );
+    }
+    this.#buffer.push(value);
+    return this.#buffer.shift();
+  }
 
-        buffer.push(chunk);
-        if (buffer.length > 2) yield buffer.shift()!;
-      }
-      if (buffer.length < 2) {
+  async *#untar(): AsyncGenerator<TarStreamEntry> {
+    for (let i = 0; i < 2; ++i) {
+      const { done, value } = await this.#reader.read();
+      if (done || value.length !== 512) {
         throw new RangeError(
           "Cannot extract the tar archive: The tarball is too small to be valid",
         );
       }
-      if (!buffer.every((value) => value.every((x) => x === 0))) {
-        throw new TypeError(
-          "Cannot extract the tar archive: The tarball has invalid ending",
-        );
-      }
-    }();
-  }
-
-  async *#untar(): AsyncGenerator<TarStreamEntry> {
+      this.#buffer.push(value);
+    }
     const decoder = new TextDecoder();
     while (true) {
       while (this.#lock) {
         await new Promise((resolve) => setTimeout(resolve, 0));
       }
 
-      const { done, value } = await this.#gen.next();
-      if (done) break;
+      // Check for premature ending
+      if (this.#buffer.every((value) => value.every((x) => x === 0))) {
+        await this.#reader.cancel("Tar Stream Finished");
```

Comment:
```suggestion
        await this.#reader.cancel(new RangeError("Tar stream finished prematurely"));
```

Reply:
This isn't meant to be an error. It's perfectly fine for it to finish with data still in the stream

Reply:
Fair enough. Let's go with `Tar stream finished prematurely` - it shouldn't be in title case and "prematurely" might provide needed context.